astropy/table/_dataframes.py

```diff
@@ -0,0 +1,517 @@
+                raise ValueError(
+                    "Cannot convert masked integer columns to DataFrame without using nullable integers. "
+                    f"Set use_nullable_int=True or remove the offending column: {n}."
+                )
+            elif old_dtype.is_float():
+                df_nw = df_nw.with_columns(nw.col(n).cast(old_dtype).alias(n))
+
+    # Convert to dataframe
+    df_native = df_nw.to_native()
+
+    # Fix pandas-like nullable integers
+    if backend_impl.is_pandas_like() and tbl.has_masked_columns and use_nullable_int:
+        for name in masked_cols:
+            dtype = array[name].dtype
+            if (dtype := array[name].dtype).kind not in INTEGER_DTYPE_KINDS:
+                continue
+
+            df_native[name] = df_native[name].astype(_numpy_to_pandas_dtype(dtype))
+
+    # Pandas-like index
+    if index:
+        df_native.set_index(index, inplace=True, drop=True)
+
+    return df_native
+
+
+def from_df(
+    df: Any, *, index: bool = False, units: Mapping[str, "UnitLike"] | None = None
+) -> "Table":
+    """Create a Table from any narwhals-compatible DataFrame."""
+    from .table import Table
+
+    if not HAS_NARWHALS:
+        raise ModuleNotFoundError(
+            "The narwhals library is required for the generic from_df method. "
+            "If you want to only convert from pandas, use the `from_pandas` method instead."
+        )
+    import narwhals as nw
+
+    # Create output
+    out = OrderedDict()
+
+    # Handle pandas index
+    if index:
+        if not hasattr(df, "index"):
+            raise ValueError(
+                "The input dataframe does not have an index. "
+                "Are you trying to convert a non-pandas dataframe? "
+                "Set `index=False` to avoid this error."
+            )
+        index_name = df.index.name or "index"
+        while index_name in df.columns:
+            index_name = "_" + index_name + "_"
+        df.reset_index(index_name, inplace=True, drop=False)
+
+    # Narwhals layer, must convert to eager
+    df_nw = nw.from_native(df)
+    if isinstance(df_nw, nw.LazyFrame):
+        df_nw = df_nw.collect()
+
+    # Handle units
+    if units is None:
+        units = {}
+    elif not isinstance(units, Mapping):
+        raise TypeError('Expected a Mapping "column-name" -> "unit"')
+    not_found = set(units.keys()) - set(df_nw.columns)
+    if not_found:
+        warnings.warn(f"`units` contains additional columns: {not_found}")
+
+    # Iterate over Narwhals columns
+    for column in df_nw.iter_columns():
+        # Unpack relevant data
+        data = column.to_numpy()
+        name = column.name
+        dtype = column.dtype
+        mask = column.is_null().to_numpy()
+        unit = units.get(name)
+
+        if isinstance(dtype, nw.Int128):
+            raise ValueError(
+                "Numpy and astropy Table do not support Int128, please use a smaller integer type."
```

Comment:
Let's avoid making statement outside our control, especially if that's redundant

```suggestion
                "Astropy Table does not support Int128, please use a smaller integer type."
```

Reply:
also, I'm not sure this problem is specific to Int128: any dtype over 64 bytes is currently not supported, as far as I know, so we'd rather check the byte size associated to the dtype instead. However, I haven't found a way to do that with narwhals' dtypes' interface. Do you know of one ? Should this also be requested upstream ?

Reply:
That is a good point. I've added a test for this error (and also a test for the lazy-only backends too). Perhaps it isn't worth having this exception for this one niche edge case and wrapping exporting to a numpy array in a try/except clause that is more generic?